notebook/notebook/notebook.py

```diff
@@ -34,40 +65,32 @@
     handlers=[fh, ch],
     level=logging.INFO)
 
-if 'BATCH_USE_KUBE_CONFIG' in os.environ:
-    kube.config.load_kube_config()
-else:
-    kube.config.load_incluster_config()
+log.info(f'AUTH_GATEWAY: {AUTH_GATEWAY}')
+log.info(f'DEFAULT_HAIL_IMAGE: {DEFAULT_HAIL_IMAGE}')
+log.info(f'KUBERNETES_TIMEOUT_IN_SECONDS: {KUBERNETES_TIMEOUT_IN_SECONDS}')
+log.info(f'INSTANCE_ID: {INSTANCE_ID}')
+
+# Prevent issues with many websockets leading to many kube connections
+# TODO: tune this, potentially watch many users with one connection
+kube.config.connection_pool_maxsize = 5000
 k8s = kube.client.CoreV1Api()
 
 app = Flask(__name__)
 sockets = Sockets(app)
 
-def read_string(f):
-    with open(f, 'r') as f:
-        return f.read().strip()
 
-KUBERNETES_TIMEOUT_IN_SECONDS = float(os.environ.get('KUBERNETES_TIMEOUT_IN_SECONDS', 5.0))
-app.secret_key = read_string('/notebook-secrets/secret-key')
-PASSWORD = read_string('/notebook-secrets/password')
-ADMIN_PASSWORD = read_string('/notebook-secrets/admin-password')
-INSTANCE_ID = uuid.uuid4().hex
+# FIXME: use hash
+def UNSAFE_user_id_transform(user_id): return user_id.replace('|', '--_--')
 
-log.info(f'KUBERNETES_TIMEOUT_IN_SECONDS {KUBERNETES_TIMEOUT_IN_SECONDS}')
-log.info(f'INSTANCE_ID {INSTANCE_ID}')
 
-try:
-    with open('notebook-worker-images', 'r') as f:
-        def get_name(line):
-            return re.search("/([^/:]+):", line).group(1)
-        WORKER_IMAGES = {get_name(line): line.strip() for line in f}
-except FileNotFoundError as e:
-    raise ValueError(
-        "working directory must contain a file called `notebook-worker-images' "
-        "containing the name of the docker image to use for worker pods.") from e
+def forbidden():
+    return 'Forbidden', 404
```

Comment:
I haven't looked over the larger code, but I just want to mention an issue: a user should not be able to probe the existence of a resource owned by another user (doesn't have list permission).  So accessing the resource of another user should give the same error (404, not found) as attempting to access a non-existent resource.

This, Forbidden seems wrong here.

Reply:
I agree with you, and it was returning 404 for every non-authorized use. I renamed the function and fixed the message.